### PR TITLE
Change password provided the actvation key is correct and has not exp…

### PR DESCRIPTION
**Pull request by fanyui**
## Description of what I changed

Change password provided the actvation key is correct and has not expired Also added test for checking if activation key is not expired or not wrong before updating  password. I also fix a bug in the method getUserByActivationKey. 

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-

## Checklist: I completed these to help reviewers :)

- [x ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/UserService.java [hunk 1]
```diff
@@ -573,4 +573,12 @@
 	 */
 	public User setUserActivationKey(User user);
 	
+	/**
+	 * Change user password given the activation key
+	 * 
+	 * @param activationKey the activation for password reset
+	 * @param newPassword the new password
+	 */
+	public void changeUserPasswordUsingActivationKey(String activationKey, String newPassword) throws APIException;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserService.java]: APIException is a runtime exception so it's pointless to add it to a throws clause

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -734,9 +734,11 @@ public User getUserByEmail(String email) throws APIException {
 	@Transactional(readOnly = true)
 	public User getUserByActivationKey(String activationKey) {
 		LoginCredential loginCred = dao.getLoginCredentialByActivationKey(activationKey);
-		String[] credTokens = loginCred.getActivationKey().split(":");
-		if(loginCred != null  && (System.currentTimeMillis() <= Long.parseLong(credTokens[1]) )) {
-			return getUser(loginCred.getUserId());
+		if (loginCred != null) {
+			String[] credTokens = loginCred.getActivationKey().split(":");
+			if (System.currentTimeMillis() <= Long.parseLong(credTokens[1])) {
+				return getUser(loginCred.getUserId());
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: I'd actually like to see a very specific exception getting thrown if the credential has expired so that the calling code know what to do next

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -760,5 +762,16 @@ public User setUserActivationKey(User user) {
...
 	}
 
+	/**
+	 * @see org.openmrs.api.UserService#changeUserPasswordUsingActivationKey(String, String);
+	 */
+	@Override
+	public void changeUserPasswordUsingActivationKey(String activationKey, String newPassword) throws APIException {
+		User user = getUserByActivationKey(activationKey);
+		if (user == null) {
+			throw new APIException("activation.key.not.correct", (Object[]) null);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: I think you should add a new exception class to throw that way the calling code knows what the issue is and take the necessary steps next. Take a look at other related password and authentication exceptions for an example.

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Why are you passing a null value in as an argument? You might as well pass in no argument

File: api/src/main/resources/messages.properties [hunk 4]
```diff
@@ -3264,6 +3264,7 @@ user.must.exist=User must exist.
 null.old.password.privilege.required=Old password is empty therefore the Edit User Passwords privilege is required.
 old.password.not.correct=Old password is not correct.
 secret.answer.not.correct=Secret answer is not correct.
+activation.key.not.correct=User activation key is not correct.
```
Comment (wluyima) [inline api/src/main/resources/messages.properties]: Error message should say is incorrect or expired

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 5]
```diff
@@ -1415,4 +1415,77 @@ public void getUserByActivationKey_shouldReturnNullIfTokenTimeExpired(){
...
+		String key = "h4ph0fpNzQCIPSw8plJI";
+		int validTime = 10 * 60 * 1000; //equivalent to 10 minutes for token to be valid
+		Long tokenTime = System.currentTimeMillis() + validTime;
+		LoginCredential credentials = dao.getLoginCredential(createdUser);
+		credentials.setActivationKey(
+		    "b071c88d6d877922e35af2e6a90dd57d37ac61143a03bb986c5f353566f3972a86ce9b2604c31a22dfa467922dcfd54fa7d18b0a7c7648d94ca3d97a88ea2fd0:"
+		            + tokenTime);
+		dao.updateLoginCredential(credentials);
+		
+		assertFalse(createdUser.hasPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS));
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: What is the use of this? A user changes their own password

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 6]
```diff
@@ -1415,4 +1415,77 @@ public void getUserByActivationKey_shouldReturnNullIfTokenTimeExpired(){
...
+		Context.authenticate(createdUser.getUsername(), "Openmr5xy");
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: What's the use of this?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 7]
```diff
@@ -1415,4 +1415,77 @@ public void getUserByActivationKey_shouldReturnNullIfTokenTimeExpired(){
...
+	@Test
+	public void changeUserPasswordUsingActivationKey_shouldNotUpdatePasswordIfActivationKeyExpired() {
+		User u = new User();
+		u.setPerson(new Person());
+		u.addName(new PersonName("Benjamin", "A", "Wolfe"));
+		u.setUsername("bwolfe");
+		u.getPerson().setGender("M");
+		User createdUser = userService.createUser(u, "Openmr5xy");
+		String key = "h4ph0fpNzQCIPSw8plJI";
+		int validTime = 10 * 60 * 1000; //equivalent to 10 minutes for token to be valid
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Is this 10 minutes?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 8]
```diff
@@ -1415,4 +1415,77 @@ public void getUserByActivationKey_shouldReturnNullIfTokenTimeExpired(){
...
+	@Test
+	public void changeUserPasswordUsingActivationKey_shouldNotUpdatePasswordIfActivationKeyIsIncorrect() {
+		User u = new User();
+		u.setPerson(new Person());
+		u.addName(new PersonName("Benjamin", "A", "Wolfe"));
+		u.setUsername("bwolfe");
+		u.getPerson().setGender("M");
+		User createdUser = userService.createUser(u, "Openmr5xy");
+		String key = "wrongactivationkeyin";
+		int validTime = 10 * 60 * 1000; //equivalent to 10 minutes for token to be valid
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: Is this 10minutes?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 7 above)
Comment (wluyima) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: Are you sure this token would have expired, this test executes and ends in less than 1min

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 3 above)
Comment (fanyui) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: This is a comparison not a parameter passing

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 7 above)
Comment (fanyui) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: yes the time is set to expire 10 minutes in the past that is why I subtracted from the line below. I thought of making a thread sleep as a resource wastage.

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 8 above)
Comment (fanyui) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java]: Yes it equivalent of 600000 milli seconds which is 10 minutes

File: api/src/main/java/org/openmrs/api/UserNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ */
+package org.openmrs.api;
+
+import org.openmrs.api.context.Context;
+
+/**
+ * Represents often fatal errors that occur when a user cannot be found due to incorrect, expired
+ * token or wrong email and or username.
+ */
+public class UserNotFoundException extends APIException {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/UserNotFoundException.java]: It's not user not found, it's invalid token

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -760,5 +763,16 @@ public User setUserActivationKey(User user) {
 		return user;
 	}
 
+	/**
+	 * @see org.openmrs.api.UserService#changeUserPasswordUsingActivationKey(String, String);
+	 */
+	@Override
+	public void changeUserPasswordUsingActivationKey(String activationKey, String newPassword) throws UserNotFoundException {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: Please I've said this before, don't add a throws clause for a runtime exception

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -760,5 +763,16 @@ public User setUserActivationKey(User user) {
...
+		User user = getUserByActivationKey(activationKey);
+		if (user == null) {
+			throw new UserNotFoundException("activation.key.not.correct", (Object[]) null);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java]: I did say remove the argument rather than passing in null

File: api/src/main/resources/messages.properties [hunk 12]
```diff
@@ -3264,6 +3264,7 @@ user.must.exist=User must exist.
 null.old.password.privilege.required=Old password is empty therefore the Edit User Passwords privilege is required.
 old.password.not.correct=Old password is not correct.
 secret.answer.not.correct=Secret answer is not correct.
+activation.key.not.correct=User activation key is incorrect or expired.
```
Comment (wluyima) [inline api/src/main/resources/messages.properties]: Rephrase this to **Invalid or expired activation key**

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 13]
```diff
@@ -1415,4 +1415,74 @@ public void getUserByActivationKey_shouldReturnNullIfTokenTimeExpired(){
...
+		Long tokenTime = System.currentTimeMillis() + validTime;
+		LoginCredential credentials = dao.getLoginCredential(createdUser);
+		credentials.setActivationKey(
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java]: For this test, you don't need to set the activation key anyways, because if the user has no activation key set then whatever value is provided is definitely incorrect

**Comment by fanyui** [comment]
@wluyima  I have updated the pull request.

File: api/src/main/java/org/openmrs/api/InvalidTokenException.java [hunk 14]
```diff
@@ -15,37 +15,27 @@
  * Represents often fatal errors that occur when a user cannot be found due to incorrect, expired
  * token or wrong email and or username.
  */
-public class UserNotFoundException extends APIException {
+public class InvalidTokenException extends APIException {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/InvalidTokenException.java]: I didn't literally  say name it InvalidTokenException, to be exact it is an Invalid activation key so it should be InvalidActivationKeyException

File: api/src/main/java/org/openmrs/api/InvalidActivationKeyException.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api;
+
+import org.openmrs.api.context.Context;
+
+/**
+ * Represents often fatal errors that occur when a user cannot be found due to incorrect, expired
+ * token or wrong email and or username.
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/InvalidActivationKeyException.java]: These javadocs seem incorrect, i thought the exception is only thrown for invalid or expired activation key

File: api/src/main/java/org/openmrs/api/InvalidActivationKeyException.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+ */
+public class InvalidActivationKeyException extends APIException {
+	
+	/**
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/InvalidActivationKeyException.java]: Why do yoou have empty comments here? Please remove them
